rfcs/1-declarative-configuration.md

```diff
@@ -0,0 +1,127 @@
+### Goals
+
+ - Documentation for all options in configuration file(s)
+ - Standard location for all configuration for tooling provided by devbase, et. al
+ - Documentation mapping 1:1 with Magefile target, with framework to consume/declare it
+ - Documentation on how to write Magefile targets, if the framework docs aren't self-explanatory enough
+ - Support for usage outside of stencil-base, et. al
+ - Not breaking, while we could do a lot more if we did a breaking release, breaking _all_ of the existing releasing tooling isn't optimal.
+
+### Non-goals
+
+ - Expose new functionality, this would be nice to have but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.
+ - Migrating `Makefile` to `Magefile`, while it'd be nice to migrate all of them over, that's also an amount of work to do. While moving to `Magefile` would be good, we shouldn't try to move everything to Go at the same time. (TL;DR: Wrap shell were needed)
+
+## Proposal
+
+The high level proposal here is generally covered in the [Summary](#summary) and [Moviation](#motiviation) sections.
+
+### Implementation Details
+
+#### Repository Configuration
+
+Configuration would live in the root of repository consuming `devbase`, in `.devbase`.
+
+Each configuration file would match the relevant `mage <target>` target, e.g.
+
+```bash
+$ ls .devbase
+build.yaml
+e2e.yaml
+docker.yaml
+```
+
+**Note**: For shared configuration, e.g. `docker` (which could be shared between `docker-build` and `docker-push`) that's also allowed and can become a unified target.
+
+#### Magefile Targets Implementation Details
+
+Each magefile target that consumes this configuration will live inside of the `devbase` repository in the `targets/<targetName>` package, which will be pulled in by `root/mage.go`
+
+The reasoning for a package per target is to make it easier to test these in isolation, as well as (generated) documentation to be localized to the packages (targets) themselves.
+
+A library will be provided at `pkg/targets`, which will provide functionality to read configuration, and other shared functions between targets.
+
+#### Magefile Targets Available Day One
+
+The following targets will be available/configurable day one:
+
+ - `build`
+ - `debug`
+ - `release`
+ - `test` (includes: `coverage` as `--coverage`, `benchmark` as `--bench`)
+ - `lint`
+ - `fmt`
+ - `e2e`
+ - `docker` (includes: `build`, `publish`)
+
+The rest will still be available for compat, but unconfigurable, e.g. `gobuild` will still build go but will not work for non-go repos.
+
+See [Examples](#examples) for an idea of what options will be available/an idea of some of the steps.
+
+#### Testing Framework
+
+Testing would be done via Go Testing, enabling us to easily write unit tests. Integration tests would, unforuantely, still need to be figured out as they rely on being ran in macOS/Linux in a predictable fashion
```

Comment:
Maybe we should consider writing this to run them all in containers, or something else that'd be more testable...